### PR TITLE
Add failing test for cache consistency bug

### PR DESCRIPTION
**Pull request by amannn**
Since the bug is really specific, I've submitted a failing test case instead of a bug report—hope this helps!

### Checklist:

- [ ] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

File: src/react/hooks/__tests__/useQuery.test.tsx [hunk 1]
```diff
@@ -3759,6 +3760,146 @@ describe('useQuery Hook', () => {
...
+            "model": "RS8",
+          },
+        },
+        {
+          "car": {
+            "__typename": "Car",
+            "make": "Audi",
+            "model": "RS8",
+          },
+        }
```
Comment (amannn) [inline src/react/hooks/__tests__/useQuery.test.tsx:3900]: Here, for the last result, `undefined` is returned instead.

**Comment by amannn** [comment]
Could someone from the team possibly have a look? We're currently experiencing this bug in a production app. Maybe @alessbell? Many thanks! 🙌

**Comment by phryneas** [comment]
@amannn sorry this took so long! You can already see a bit of progress in #11078.
I'll make sure that that PR will have an installable npm package version for you so you can already use this while I go over some potential edge cases.

**Comment by amannn** [comment]
@phryneas Thank you so much for the update! I think I'll wait for the stable release, but if it helps, I can test a prerelease in my app. We have quite a large test suite, sometimes this uncovers the effects of subtle changes in Apollo Client!

**Comment by phryneas** [comment]
Any feedback you can give here is very welcome! :)

**Comment by phryneas** [comment]
On a general note: we are in a code freeze right now until we have released Apollo Client 3.8 (currently at RC1), so this will probably be in one of the first patches to hit after the stable release of that.

**Comment by amannn** [comment]
@phryneas I just tried the prerelease in my app and that fixes the regression test I created for this issue! 🎉 Also everything else is passing, looks great! 🙌

**Comment by amannn** [comment]
> On a general note: we are in a code freeze right now until we have released Apollo Client 3.8 (currently at RC1), so this will probably be in one of the first patches to hit after the stable release of that.

@phryneas Is there more work / an official release with the fix from #11078 planned? Thank you!

**Comment by phryneas** [comment]
Hi @amannn!
I'm very sorry, this somehow disappeared from my list. I'll try to get back to it at some point during this week - thank you for the ping.

**Comment by amannn** [comment]
@phryneas Sorry for bothering you again, is this bug currently a priority? The behavior described in the test case in this PR unfortunately causes random failures in a production app. It would be really helpful if Apollo Client would be resilient to such race conditions. Thank you for your help!

**Comment by phryneas** [comment]
I have it on my mind, and I can prove it 😅 
I sent this chat message to our channel internally yesterday:
image

=> We're close to getting to AC 3.9, and right now I'm hard working on the last few features that we need in place so we can put out a beta (since they have an impact on the external api). 
Once that's done, this is one of the last few bug fixes I intend to land - I'm prioritizing other stuff right now though, since it's blocking the "beta" label for us.

I'm really sorry for the delay, but I'm juggling a lot of priorities right now.

**Comment by amannn** [comment]
Thank you so much for the update! Sure, I understand. Thank you for all the work you put into Apollo Client!

**Comment by phryneas** [comment]
The fix has been merged and will make it into the next AC 3.9 beta :) I'm gonna close this one out now - thanks a ton again for the reproduction!

**Comment by amannn** [comment]
That's great to hear, thank you so much for your help! 🙌
